src/wallet/signer.ts

```diff
@@ -0,0 +1,222 @@
+import { BigNumber } from "bignumber.js";
+import { flatMap } from "lodash";
+import { decodeAccountID } from "ripple-address-codec";
+import {
+  encode,
+  decode,
+  encodeForSigning,
+  encodeForSigningClaim,
+} from "ripple-binary-codec";
+import { JsonObject } from "ripple-binary-codec/dist/types/serialized-type";
+import {
+  sign as signWithKeypair,
+  verify as verifySignature,
+} from "ripple-keypairs";
+
+import { ValidationError } from "../common/errors";
+import { SignedTransaction } from "../common/types/objects";
+import { Signer } from "../models/common";
+import { Transaction } from "../models/transactions";
+import { verifyBaseTransaction } from "../models/transactions/common";
+import { computeBinaryTransactionHash } from "../utils";
+import Wallet from "../Wallet";
+
+/**
+ * Uses a wallet to cryptographically sign a transaction which proves the owner of the wallet
+ * is issuing this transaction.
+ *
+ * @param wallet - A Wallet that holds your cryptographic keys.
+ * @param tx - The Transaction that is being signed.
+ * @returns A signed Transaction and a Transaction id that corresponds to the pre-signed Transaction.
+ */
+function sign(wallet: Wallet, tx: Transaction): SignedTransaction {
+  return wallet.signTransaction(tx, { signAs: "" });
+}
+
+/**
+ * Takes several transactions (in object or blob form) and creates a single transaction with all Signers
+ * that then gets signed and returned.
+ *
+ * @param transactions - An array of Transactions (in object or blob form) to combine and sign.
+ * @returns A single signed Transaction which has all Signers from transactions within it.
+ * @throws ValidationError if:
+ * - There were no transactions given to sign
+ * - The SigningPubKey field is not the empty string in any given transaction
+ * - Any transaction is missing a Signers field.
+ */
+function multisign(
+  transactions: Array<Transaction | string>
+): SignedTransaction {
+  if (transactions.length === 0) {
+    throw new ValidationError("There were 0 transactions given to multisign");
+  }
+
+  transactions.forEach((txOrBlob) => {
+    const tx: Transaction = getDecodedTransaction(txOrBlob);
+
+    // This will throw a more clear error for JS users if any of the supplied transactions has incorrect formatting
+    // TODO: Replace this with verify() (The general validation function for all Transactions)
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- verify does not accept Transaction type
+    verifyBaseTransaction(tx as unknown as Record<string, unknown>);
+
+    if (tx.SigningPubKey !== "") {
+      throw new ValidationError(
+        "For multisigning the transaction must include the SigningPubKey field as an empty string."
+      );
+    }
+
+    if (tx.Signers == null) {
+      throw new ValidationError(
+        "For multisigning the transaction must include a Signers field containing an array of signatures."
+      );
+    }
```

Comment:
Don't the transactions need to be signed individually for them to be combined?

Reply:
This actually sparked a lot of confusion for me and after talking to Nathan just now we realized we were missing a piece of the multisign workflow:

- Combining multisigned transactions requires them to be signed individually in a special format. In Rippled this is done by the "sign_for" request. Specifically, sign_for creates the "Signers" field instead of changing "TxnSignature" or "SigningPubKey" fields. (We didn't want to implement "sign_for" because it signs over the internet, but we do still need a way to sign using "Signers" instead of "TxnSignature" on the transaction)
- So, we need an offline 'multisign' function takes a transaction and signs it in the multsign format
- And we need a combineMultisigned function (which is what we were previously calling 'multisign' in the Signers ticket or 'combine' in 1.x)

Reply:
The current names I'm using are 'multisign' and 'combineMultisigned' for these two functions, @mvadari what are your thoughts on those names?

Reply:
Those names feel confusing to me. But could we instead have one multisign function that handles both parts? The 3 parts of a `Signer` object are pretty easy to derive in the same method as where we're combining stuff.

Reply:
Potentially, the issue would be that only one of the many signers would be able to use the combined version. The default workflow would be A multisigns, B multisigns, C multisigns, then those three are combined. So only by the time C signs do we have all the information needed for combine. 

If all signatures needed for multisigning were owned by the same node, we could simplify it to sign all in the combine method, but I think there are important use cases where you need signatures from independent networks (which sign offline on their own networks and share the end result). 

Maybe signForMultisign and combineMultisigned would be better names?

Reply:
I think you could still take in the existing structure - the `Signer` object needs `Account`, `SignerPublicKey`, and `TxnSignature`, which are all in a signed transaction anyways. 
Hm, I guess you still need to sign it differently with ripple-binary-codec, so you'd need to have that helper method anyways.
I don't like the `for` part of `signForMultisigned` - maybe just `signMultisign`?